[VPlan] Remove hoisting to preheader in VPTransformState::get

lukel97 opened the pull request:
Now that the start values of reductions are explicitly modelled as a VPInstruction in 6108d50, we no longer really need to hoist when executing. 

The last remaining use of this is one `VPPredInstPHIRecipe` test case

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -290,18 +290,22 @@ Value *VPTransformState::get(const VPValue *Def, bool NeedsScalar) {
 
   auto GetBroadcastInstrs = [this, Def](Value *V) {
+    VPBasicBlock *PreheaderVPBB = nullptr;
+    if (auto *Header = vputils::getFirstLoopHeader(*Plan, VPDT))
```
At this stage during execution the regions will have been dissolved, except for replicator regions. But there's a check for R->isReplicator(), so I think `VPBlockUtils::isHeader` should only return true if there's 2 predecessors

fhahn commented on llvm/test/Transforms/LoopVectorize/X86/pr141968-instsimplifyfolder.ll:
```diff
@@ -97,8 +99,6 @@ define i8 @pr141968(i1 %cond, i8 %v) {
 ; CHECK:       [[PRED_SDIV_CONTINUE30]]:
-; CHECK-NEXT:    [[BROADCAST_SPLATINSERT31:%.*]] = insertelement <16 x i8> poison, i8 [[V]], i64 0
-; CHECK-NEXT:    [[BROADCAST_SPLAT32:%.*]] = shufflevector <16 x i8> [[BROADCAST_SPLATINSERT31]], <16 x i8> poison, <16 x i32> zeroinitializer
 ; CHECK-NEXT:    [[PREDPHI:%.*]] = select <16 x i1> [[BROADCAST_SPLAT]], <16 x i8> zeroinitializer, <16 x i8> [[BROADCAST_SPLAT32]]
```
Hm, is the only user a VPPredInstPHIRecipe? It uses the vector value, but has `usesScalars` returns true. Should it return false?

lukel97 replied:
Good catch, removed the overridden definition in ba9e50319f749572f4cb42450ac2c23c5bbc68fa so it should return false now.

fhahn replied:
Are the changes  to `GetBroadcastInstrs` still needed? It sounds like the `SafeToHoist` could be removed now? I *think* the last user of the code has been removed in 6108d50aed6190aa094c168f96899adc5621cc89.

If that's the case would be good to have one patch updating VPPredInstPHIRecipe and one stripping the SafeToHoist logic.

lukel97 replied:
I've opened https://github.com/llvm/llvm-project/pull/143798 for VPPredInstPHIRecipe, and reworked this PR to remove the SafeToHoist logic. 

In either case we still end up with the test diff in this test case, i.e. changing usesScalars doesn't seem to affect it.

lukel97 committed:
[VPlan] Remove hoisting to preheader in VPTransformState::get

Now that the start values of reductions are explicitly modelled as a VPInstruction in 6108d50, we no longer really need to hoist when executing.

From the discussion in https://github.com/llvm/llvm-project/pull/142594#discussion_r2140339682
